src/main/java/org/truffleruby/parser/BodyTranslator.java

```diff
@@ -2450,51 +2450,84 @@ public RubyNode visitOpElementAsgnNode(OpElementAsgnParseNode node) {
+
+            return block(node, writeReceiverToTemp, write);
         }
+    }
 
-        copyNewline(node, operation);
+    private RubyNode block(OpElementAsgnParseNode node, ParseNode writeReceiverToTemp, ParseNode main) {
+        final BlockParseNode block = new BlockParseNode(node.getPosition());
+        block.add(writeReceiverToTemp);
+        block.add(main);
 
-        final ParseNode writeArgs;
+        final RubyNode ret = block.accept(this);
+        return addNewlineIfNeeded(node, ret);
+    }
+
+    private ParseNode write(
+            OpElementAsgnParseNode node,
+            ParseNode readReceiverFromTemp,
+            ParseNode value) {
+
+        final ParseNode readArguments = node.getArgsNode();
+        final ParseNode writeArguments;
         // Like ParserSupport#arg_add, but copy the first node
-        if (readArgs instanceof ArrayParseNode) {
+        if (readArguments instanceof ArrayParseNode) {
             final ArrayParseNode readArgsCopy = new ArrayParseNode(node.getPosition());
-            readArgsCopy.addAll((ArrayParseNode) readArgs).add(operation);
-            writeArgs = readArgsCopy;
+            readArgsCopy.addAll((ArrayParseNode) readArguments).add(value);
+            writeArguments = readArgsCopy;
         } else {
-            writeArgs = new ArgsPushParseNode(node.getPosition(), readArgs, operation);
+            writeArguments = new ArgsPushParseNode(node.getPosition(), readArguments, value);
         }
 
-        final ParseNode arrayWrite = new CallParseNode(node.getPosition(), readArrayFromTemp, "[]=", writeArgs, null);
+        return new CallParseNode(node.getPosition(), readReceiverFromTemp, "[]=", writeArguments, null);
+    }
 
-        final BlockParseNode block = new BlockParseNode(node.getPosition());
-        block.add(writeArrayToTemp);
-        block.add(arrayWrite);
+    private ParseNode operation(
+            OpElementAsgnParseNode node,
+            ParseNode readReceiverFromTemp,
+            String op,
+            ParseNode right) {
 
-        final RubyNode ret = block.accept(this);
-        return addNewlineIfNeeded(node, ret);
+        final ParseNode read = new CallParseNode(
+                node.getPosition(), readReceiverFromTemp, "[]", node.getArgsNode(), null);
+        ParseNode operation;
+        switch (op) {
+            case "||":
+                operation = new OrParseNode(node.getPosition(), read, right);
+                break;
+            case "&&":
+                operation = new AndParseNode(node.getPosition(), read, right);
+                break;
+            default:
+                operation = new CallParseNode(
+                        node.getPosition(),
+                        read,
+                        node.getOperatorName(),
+                        buildArrayNode(node.getPosition(), right),
+                        null);
+                break;
+        }
+
+        copyNewline(node, operation);
+        return operation;
```

Comment:
This logic needs a spec or removing a tag if there is already a spec.

Reply:
I am making a note and I'll check or add later. I need this merged asap for 33